fastnml/reader.py

```diff
@@ -0,0 +1,251 @@
+                    # convert the string to a Python value:
+                    value = _nml_value_to_python_value(d[1].rstrip(', '))
+
+                    # add this value to the namelist:
+                    path = d[0].strip()
+                    _pathSet(nml[namelist_name], path, value)
+        except Exception:
+            nml = None
+    if nml is None:
+        try:
+            nml = parser.reads(''.join(lines))  # f90nml 1.1
+        except AttributeError:
+            nml = parser._readstream(iter(lines), {})  # previous ver
+    return nml
+
+
+# def split_namelist_str(s: str):
+#     """alternate version of split_namelist_file with a string as an input"""
+#     namelists = []
+#     i = -1
+#     started = False
+#     f = s.split('\n')
+#     for line in f:
+#         line = line.strip()
+#         if (len(line) > 0):
+#             if (line[0] != '!'):
+#                 if (line[0] == '&'):  # start a namelist
+#                     i = i + 1
+#                     namelists.append([])
+#                     started = True
+#                 elif (line[0:1] == '/'):  # end a namelist
+#                     started = False
+#                     namelists[i].append(line)
+#                     continue
+#                 if started:
+#                     namelists[i].append(line)
+#     return namelists
+
+
+def _split_namelist_file(filename: str) -> List[str]:
+    """ split a namelist file into an array of namelist strings """
+    namelists = list()
+    i = -1
+    started = False
+    with open(filename, 'r') as f:
+        for line in f:
+            line = line.strip()
+            if (len(line) > 0):
+                if (line[0] != '!'):
+                    if (line[0] == '&'):  # start a namelist
+                        i = i + 1
+                        namelists.append(list())
+                        started = True
+                    elif (line[0:1] == '/'):  # end a namelist
+                        started = False
+                        namelists[i].append(line)
+                        continue
+                    if started:
+                        namelists[i].append(line)
+
+    return namelists
+
+
+def read_namelist(filename: str, *, n_threads: int = 4,
+                  parser: Parser = None,
+                  simple: bool = True) -> Namelist:
+    """
+    Read a namelist quickly.
+
+    For non-threaded use, set `n_threads=0`.
+    """
+    nml = Namelist({})
+
+    def _loop_over_results(r):
+        for key, value in r.items():
+            if key in nml:
+                # array of namelists:
+                if isinstance(nml[key], list):
+                    nml[key].append(value)
+                else:
+                    nml[key] = [nml[key], value]
+            else:
+                nml[key] = value
+
+    if not parser:
+        parser = Parser()
+
+    namelists = _split_namelist_file(filename)
+    results = list()
+    results_append = results.append
```

Comment:
why do this?

Reply:
setting `results_append = results.append`?  It provides a shortcut to the exact method on the object so it does not need to resolve the dot operator every iteration through the loop.  Provides a very slight increase in speed as a result.